substrate/client/chain-spec/src/lib.rs

```diff
@@ -16,19 +16,229 @@
+//!       </td>
+//!       <td>A JSON object with two fields: <code>top</code> and <code>children_default</code>.
+//! Each field is a map of <code>key => value</code> pairs representing entries in a genesis storage
+//! trie.</td>
+//!     </tr>
+//!   </tbody>
+//! </table>
+//!
+//! For production or long-lasting blockchains, using the `raw` format in the chain specification is
+//! recommended.
+//!
+//! JSON examples in the [following section](#json-chain-specification-example) illustrate the `raw`
+//! and `runtimeGenesisConfigPatch` genesis fields.
+//!
+//! # From Initial State to Raw Genesis.
+//!
+//! To generate a raw genesis storage from the JSON representation of the runtime genesis config,
+//! the node needs to interact with the runtime.
+//!
+//! This interaction involves passing the runtime genesis config JSON blob to the runtime using the
+//! [`sp_genesis_builder::GenesisBuilder::build_config`] function. During this operation, the
+//! runtime converts the JSON representation of the genesis config into [`sp_io::storage`] items. It
+//! is a crucial step for computing the storage root hash, which is a key component in determining
+//! the genesis hash.
+//!
+//! Consequently, the runtime must support the [`sp_genesis_builder::GenesisBuilder`] API to
+//! utilize either [`RuntimeGenesisConfigPatch`][patch] or [`RuntimeGenesisConfig`][full] formats.
+//!
+//! [patch]: chain_spec::Genesis<G>::RuntimeGenesisConfigPatch
+//! [full]: chain_spec::Genesis<G>::RuntimeGenesisConfig
+//!
+//! This entire process is encapsulated within the implementation of the [`BuildStorage`] trait,
+//! which can be accessed through the [`ChainSpec::as_storage_builder`] method. There is an
+//! intermediate internal helper that facilitates this interaction,
+//! [`GenesisConfigBuilderRuntimeCaller`], which serves as a straightforward wrapper for
+//! [`sc_executor::WasmExecutor`].
+//!
+//! In case of `raw` genesis state the node does not interact with the runtime regarding the
+//! computation of initial state.
+//!
+//! The plain and `raw` chain specification JSON blobs can be found in
+//! [JSON examples](#json-chain-specification-example) section.
+//!
+//! # Optional Code Mapping
+//! Optional map of `block_number` to `wasm_code`.
+//!
+//! The given `wasm_code` will be used to substitute the on-chain wasm code starting with the
+//! given block number until the `spec_version` on-chain changes. The given `wasm_code` should
+//! be as close as possible to the on-chain wasm code. A substitute should be used to fix a bug
+//! that cannot be fixed with a runtime upgrade, if for example the runtime is constantly
+//! panicking. Introducing new runtime APIs isn't supported, because the node
+//! will read the runtime version from the on-chain wasm code. Use this functionality only when
+//! there is no other way around it, and only patch the problematic bug; the rest should be done
+//! with an on-chain runtime upgrade.
```

Comment:
I'd insert some empty lines here and there to create some shorter paragraphs.

For example "
...version from the on-chain wasm code.

Use this functionality only when ....